llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -511,12 +604,40 @@ void validateUsedAspectsForFunctions(const FunctionToAspectsMapTy &Map,
   }
 }
 
+struct FunctionsClassifiedByTheirFunctionCall {
+  FunctionSetTy CalledUnconditionally;
+  FunctionSetTy CalledConditionally;
+};
+
+void addAllCalledFunctionsRecursivelyToFunctionsClassifiedByTheirFunctionCallStruct(
+    Function *F, const CallGraphTy &CG,
+    FunctionsClassifiedByTheirFunctionCall &Funcs,
+    SmallPtrSet<const Function *, 16> &Visited) {
+  const auto It = CG.find(F);
+  if (It == CG.end())
+    return;
+  for (Function *Callee : It->second) {
+    if (!Callee->hasFnAttribute("sycl-call-if-on-device-conditionally") &&
+        Funcs.CalledUnconditionally.contains(F)) {
+      Funcs.CalledUnconditionally.insert(Callee);
+    }
+    if (Callee->hasFnAttribute("sycl-call-if-on-device-conditionally") ||
```

Comment:
To avoid repeating the attribute check, I suggest to do it once and store result in a `bool` variable